File: mcp_visualizer/poc/claude_client.py
"""Claude API client with MCP tool schema conversion."""

import json
import sys
from typing import Any

from anthropic import Anthropic


class ClaudeClient:
    """Client for interacting with Claude API with MCP tool support."""

    def __init__(self, api_key: str):
        """Initialize the Claude client.

        Args:
            api_key: Anthropic API key
        """
        self.client = Anthropic(api_key=api_key)
        self.model = "claude-sonnet-4-20250514"

    def convert_mcp_tools_to_claude_format(
        self, mcp_tools: list[dict[str, Any]]
    ) -> list[dict[str, Any]]:
        """Convert MCP tool schemas to Claude API tool format.

        Args:
            mcp_tools: List of MCP tool definitions

        Returns:
            List of Claude API tool definitions
        """
        print("\n[HOST] Converting MCP tool schemas to Claude API format...", file=sys.stderr)

        claude_tools = []
        for mcp_tool in mcp_tools:
            claude_tool = {
                "name": mcp_tool["name"],
                "description": mcp_tool.get("description", ""),
                "input_schema": mcp_tool.get("inputSchema", {"type": "object", "properties": {}}),
            }
            claude_tools.append(claude_tool)

            print(f"[HOST] Converted tool: {mcp_tool['name']}", file=sys.stderr)

        return claude_tools

    def call_claude(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
        max_tokens: int = 1024,
    ) -> dict[str, Any]:
        """Call Claude API with messages and optional tools.

        Args:
            messages: Conversation messages
            tools: Optional list of tools in Claude format
            max_tokens: Maximum tokens for response

        Returns:
            Claude API response as dict
        """
        kwargs: dict[str, Any] = {
            "model": self.model,
            "max_tokens": max_tokens,
            "messages": messages,
        }

        if tools:
            kwargs["tools"] = tools

        print(f"\n[CLAUDE] Sending request to Claude API ({self.model})", file=sys.stderr)
        print(f"[CLAUDE] Message count: {len(messages)}", file=sys.stderr)
        if tools:
            print(f"[CLAUDE] Available tools: {[t['name'] for t in tools]}", file=sys.stderr)

        response = self.client.messages.create(**kwargs)

        response_dict = {
            "id": response.id,
            "type": response.type,
            "role": response.role,
            "content": [block.model_dump() for block in response.content],
            "model": response.model,
            "stop_reason": response.stop_reason,
            "usage": response.usage.model_dump(),
        }

        print(f"[CLAUDE] Response received:", file=sys.stderr)
        print(f"[CLAUDE] Stop reason: {response.stop_reason}", file=sys.stderr)
        print(f"[CLAUDE] Content blocks: {len(response.content)}", file=sys.stderr)

        # Log content blocks
        for i, block in enumerate(response.content):
            if block.type == "text":
                print(f"[CLAUDE] Block {i}: text ({len(block.text)} chars)", file=sys.stderr)
            elif block.type == "tool_use":
                print(f"[CLAUDE] Block {i}: tool_use - {block.name}", file=sys.stderr)
                print(f"[CLAUDE]   Input: {json.dumps(block.input, indent=2)}", file=sys.stderr)

        return response_dict

    def extract_tool_calls(self, response: dict[str, Any]) -> list[dict[str, Any]]:
        """Extract tool calls from Claude response.

        Args:
            response: Claude API response

        Returns:
            List of tool calls with id, name, and input
        """
        tool_calls = []
        for block in response["content"]:
            if block["type"] == "tool_use":
                tool_calls.append(
                    {
                        "id": block["id"],
                        "name": block["name"],
                        "input": block["input"],
                    }
                )
        return tool_calls

    def format_tool_result_for_claude(
        self, tool_call_id: str, tool_result: dict[str, Any]
    ) -> dict[str, Any]:
        """Format MCP tool result for Claude API.

        Args:
            tool_call_id: The tool call ID from Claude's tool_use block
            tool_result: The result from MCP server

        Returns:
            Formatted tool result for Claude
        """
        # Extract text content from MCP result
        content_text = ""
        if "content" in tool_result and tool_result["content"]:
            for content_item in tool_result["content"]:
                if content_item.get("type") == "text":
                    content_text += content_item.get("text", "")

        return {
            "type": "tool_result",
            "tool_use_id": tool_call_id,
            "content": content_text or json.dumps(tool_result),
        }

File: mcp-inspector-app/lib/llm/claude-client.ts
/**
 * Claude API Client
 *
 * Provides two-phase inference pattern:
 * 1. Planning inference - LLM selects tools to use
 * 2. Synthesis inference - LLM generates final response with tool results
 *
 * Based on: POC claude_client.py
 */

import Anthropic from '@anthropic-ai/sdk';
import type {
  ClaudeTool,
  ClaudeToolUse,
  ClaudeToolResult
} from './tool-formatter';
import {
  extractToolCalls,
  extractTextResponse,
  buildConversationWithResults
} from './tool-formatter';

/**
 * Claude client configuration
 */
export interface ClaudeClientConfig {
  apiKey: string;
  model?: string;
  maxTokens?: number;
}

/**
 * Inference context for event recording
 */
export interface InferenceContext {
  phase: 'planning' | 'synthesis';
  userMessage: string;
  tools: ClaudeTool[];
}

/**
 * Inference result
 */
export interface InferenceResult {
  message: Anthropic.Message;
  toolCalls: ClaudeToolUse[];
  textResponse: string;
  usage: {
    inputTokens: number;
    outputTokens: number;
  };
}

/**
 * Claude API Client
 *
 * Handles communication with Claude API for two-phase inference.
 */
export class ClaudeClient {
  private client: Anthropic;
  private model: string;
  private maxTokens: number;

  constructor(config: ClaudeClientConfig) {
    this.client = new Anthropic({ apiKey: config.apiKey });
    this.model = config.model || 'claude-sonnet-4-20250514';
    this.maxTokens = config.maxTokens || 1024;
  }

  /**
   * First inference: Planning (tool selection)
   *
   * LLM analyzes user query and selects which tools to use.
   * Returns tool_use blocks if tools are needed, or text response if not.
   *
   * @param userMessage - User's query
   * @param tools - Available tools (Claude format)
   * @returns Inference result with tool calls or text
   */
  async planningInference(
    userMessage: string,
    tools: ClaudeTool[]
  ): Promise<InferenceResult> {
    const startTime = Date.now();

    console.log('[ClaudeClient] Planning inference started');
    console.log(`[ClaudeClient] User message: "${userMessage}"`);
    console.log(`[ClaudeClient] Available tools: ${tools.length}`);

    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: this.maxTokens,
      messages: [{ role: 'user', content: userMessage }],
      tools,
    });

    const processingTime = Date.now() - startTime;

    console.log(`[ClaudeClient] Planning inference completed in ${processingTime}ms`);
    console.log(`[ClaudeClient] Stop reason: ${response.stop_reason}`);
    console.log(`[ClaudeClient] Content blocks: ${response.content.length}`);

    // Extract tool calls and text
    const toolCalls = extractToolCalls(response.content);
    const textResponse = extractTextResponse(response.content);

    console.log(`[ClaudeClient] Tool calls: ${toolCalls.length}`);
    if (toolCalls.length > 0) {
      toolCalls.forEach((tc, i) => {
        console.log(`[ClaudeClient]   ${i + 1}. ${tc.name}`);
      });
    }

    return {
      message: response,
      toolCalls,
      textResponse,
      usage: {
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
      }
    };
  }

  /**
   * Second inference: Synthesis (final response)
   *
   * LLM generates natural language response using tool results.
   * Requires conversation history with tool_use and tool_result blocks.
   *
   * @param userMessage - Original user query
   * @param planningContent - Content from planning response (includes tool_use)
   * @param toolResults - Results from tool executions
   * @param tools - Available tools (same as planning)
   * @returns Inference result with final text response
   */
  async synthesisInference(
    userMessage: string,
    planningContent: any[],
    toolResults: ClaudeToolResult[],
    tools: ClaudeTool[]
  ): Promise<InferenceResult> {
    const startTime = Date.now();

    console.log('[ClaudeClient] Synthesis inference started');
    console.log(`[ClaudeClient] Tool results: ${toolResults.length}`);

    // Build conversation history
    const messages = buildConversationWithResults(
      userMessage,
      planningContent,
      toolResults
    );

    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: this.maxTokens,
      messages,
      tools,
    });

    const processingTime = Date.now() - startTime;

    console.log(`[ClaudeClient] Synthesis inference completed in ${processingTime}ms`);
    console.log(`[ClaudeClient] Stop reason: ${response.stop_reason}`);
    console.log(`[ClaudeClient] Content blocks: ${response.content.length}`);

    // Extract text (should not have tool_use blocks in synthesis)
    const textResponse = extractTextResponse(response.content);
    const toolCalls = extractToolCalls(response.content); // Should be empty

    if (toolCalls.length > 0) {
      console.warn(`[ClaudeClient] WARNING: Synthesis returned tool calls (unexpected)`);
    }

    return {
      message: response,
      toolCalls,
      textResponse,
      usage: {
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
      }
    };
  }

  /**
   * Get current model name
   */
  getModel(): string {
    return this.model;
  }

  /**
   * Get max tokens setting
   */
  getMaxTokens(): number {
    return this.maxTokens;
  }
}

/**
 * Create Claude client instance with environment configuration
 */
export function createClaudeClient(apiKey?: string): ClaudeClient {
  const key = apiKey || process.env.ANTHROPIC_API_KEY;

  if (!key) {
    throw new Error(
      'ANTHROPIC_API_KEY not found. Set it in .env.local or pass to createClaudeClient()'
    );
  }

  return new ClaudeClient({
    apiKey: key,
    model: 'claude-sonnet-4-20250514',
    maxTokens: 1024,
  });
}
